rpcs3/Emu/Cell/Modules/cellImeJp.cpp

```diff
@@ -1,283 +1,1067 @@
-	CELL_IMEJP_ERROR_PARAM              = 0x8002bf41,
-	CELL_IMEJP_ERROR_IME_ALREADY_IN_USE = 0x8002bf51,
-	CELL_IMEJP_ERROR_OTHER              = 0x8002bfff,
-};
+	format_enum(out, arg, [](auto error)
+	{
+		switch (error)
+		{
+			STR_CASE(CELL_IMEJP_ERROR_ERR);
+			STR_CASE(CELL_IMEJP_ERROR_CONTEXT);
+			STR_CASE(CELL_IMEJP_ERROR_ALREADY_OPEN);
+			STR_CASE(CELL_IMEJP_ERROR_DIC_OPEN);
+			STR_CASE(CELL_IMEJP_ERROR_PARAM);
+			STR_CASE(CELL_IMEJP_ERROR_IME_ALREADY_IN_USE);
+			STR_CASE(CELL_IMEJP_ERROR_OTHER);
+		}
 
-static uint16_t s_ime_string[256];
+		return unknown;
+	});
+}
+
+using sys_memory_container_t = u32;
+
+ime_jp_manager::ime_jp_manager()
+{
+	if ((s32)g_ps3_process_info.sdk_ver < 0x360000) // firmware < 3.6.0
+		allowed_extensions = CELL_IMEJP_EXTENSIONCH_UD85TO94 | CELL_IMEJP_EXTENSIONCH_OUTJIS;
+	else
+		allowed_extensions = CELL_IMEJP_EXTENSIONCH_UD09TO15 | CELL_IMEJP_EXTENSIONCH_UD85TO94 | CELL_IMEJP_EXTENSIONCH_OUTJIS;
+}
+
+bool ime_jp_manager::check_handle(CellImeJpHandle handle)
+{
+	// TODO
+	return !!handle;
+}
+
+bool ime_jp_manager::addChar(u16 c)
+{
+	if (!c || cursor >= (CELL_IMEJP_STRING_MAXLENGTH - 1) || cursor > input_string.length())
+		return false;
+
+	std::u16string tmp;
+	tmp += c;
+	input_string.insert(cursor++, tmp);
+	cursor_end = cursor;
+	input_state = CELL_IMEJP_BEFORE_CONVERT;
+	return true;
+}
+
+bool ime_jp_manager::addString(vm::cptr<u16> str)
+{
+	if (!str)
+		return false;
+
+	for (size_t i = 0; i < CELL_IMEJP_STRING_MAXLENGTH; i++)
+	{
+		if (!addChar(str[i]))
+			return false;
+	}
+
+	return true;
+}
+
+bool ime_jp_manager::backspaceWord()
+{
+	if (!cursor || cursor > CELL_IMEJP_STRING_MAXLENGTH || cursor > input_string.length())
+		return false;
+
+	input_string.erase(--cursor);
+	cursor_end = cursor;
+
+	if (input_string.empty())
+		input_state = CELL_IMEJP_BEFORE_INPUT;
+
+	return true;
+}
+
+bool ime_jp_manager::deleteWord()
+{
+	if (cursor >= (CELL_IMEJP_STRING_MAXLENGTH - 1) || cursor > (input_string.length() - 1))
+		return false;
+
+	input_string.erase(cursor);
+	cursor_end = cursor;
+
+	if (input_string.empty())
+		input_state = CELL_IMEJP_BEFORE_INPUT;
+
+	return true;
+}
+
+void ime_jp_manager::moveCursor(s8 amount)
+{
+	if (amount > 0)
+	{
+		cursor = std::min(cursor + amount, input_string.length() - 1);
+		cursor_end = std::min(cursor_end + amount, input_string.length() - 1);
+	}
+	else if (amount < 0)
+	{
+		cursor = std::max(0, (s32)cursor + amount);
+		cursor_end = std::max((s32)cursor, (s32)cursor_end + amount);
+	}
+}
+
+void ime_jp_manager::moveCursorEnd(s8 amount)
+{
+	if (amount > 0)
+	{
+		cursor_end = std::max((s32)cursor, (s32)std::min(cursor_end + amount, input_string.length() - 1));
+	}
+	else if (amount < 0)
+	{
+		cursor_end = std::max((s32)cursor, (s32)cursor_end + amount);
+	}
+}
 
-s32 cellImeJpOpen()
+error_code cellImeJpOpen(sys_memory_container_t container_id, vm::ptr<CellImeJpHandle> hImeJpHandle, vm::cptr<CellImeJpAddDic> addDicPath)
```

Comment:
Use `vm::ptr<be_t<CellImeJpHandle>>` or `vm::pptr<void>` to workaround endianess problem.